worker2/worker.go

```diff
@@ -0,0 +1,206 @@
+	pollingInterval       time.Duration
+	reclaimOffset         time.Duration
+	concurrency           int
+	// State
+	running        atomics.Bool
+	activeTasks    atomics.Counter
+	stopNow        atomics.Barrier
+	stopGracefully atomics.Barrier
+}
+
+// New creates a new Worker
+func New(options Options) (*Worker, error) {
+	w := &Worker{}
+
+	return w, nil
+}
+
+// PayloadSchema returns the schema for task.payload
+func (w *Worker) PayloadSchema() schematypes.Schema {
+	return nil
+}
+
+// Start process tasks, returns nil if stopped gracefully
+func (w *Worker) Start() error {
+	// Ensure that we don't start running twice
+	if w.running.Swap(true) {
+		panic("Worker.Start() called while worker is already running")
+	}
+	defer w.running.Set(false)
+
+	// Reset barriers
+	w.stopNow = atomics.Barrier{}
+	w.stopGracefully = atomics.Barrier{}
+	w.stopNow.Forward(w.stopGracefully.Fall) // stopNow implies stopGracefully
+	defer w.stopNow.Fall()                   // Always lower the barrier
+
+	// Create LifeCyclePolicy
+	lifecycle := lifecyclepolicy.New(lifecyclepolicy.Options{
+		Monitor: w.environment.Monitor.WithPrefix("life-cycle-policy"),
+		Config:  w.lifeCyclePolicyConfig,
+		Worker:  w,
+	})
+
+	// Track idle time
+	idleTimer := atomics.StopWatch{}
+	go func() {
+		for {
+			select {
+			case <-w.activeTasks.Changed():
+				if w.activeTasks.Value() > 0 {
+					idleTimer.Reset() // Reset also stop the timer
+				} else {
+					idleTimer.Start()
+				}
+			case <-w.stopGracefully.Barrier():
+				break
+			}
+		}
+	}()
+
+	for !w.stopGracefully.IsFallen() {
+		// Claim tasks
+		claims, err := w.queue.ClaimWork(w.provisionerID, w.workerType, &queue.ClaimWorkRequest{
+			WorkerGroup: w.workerGroup,
+			WorkerID:    w.workerID,
+			Tasks:       w.concurrency - w.activeTasks.Value(),
+		})
+		if err != nil {
+			w.monitor.ReportError(err, "failed to ClaimWork")
+			lifecycle.ReportNonFatalError()
+		}
+
+		// If we have claims we MUST always handle, even if we have stopNow!
+		if claims != nil {
+			for _, claim := range claims.Tasks {
+				// Start processing tasks
+				w.activeTasks.Add(1)
+				go w.processClaim(taskClaim{
+					TaskID: claim.Status.TaskID,
+					RunID:  claim.RunID,
+					Task:   claim.Task,
+					Credentials: tcclient.Credentials{
+						ClientID:    claim.Credentials.ClientID,
+						AccessToken: claim.Credentials.AccessToken,
+						Certificate: claim.Credentials.Certificate,
+					},
+					TakenUntil: time.Time(claim.TakenUntil),
+				}, lifecycle)
+			}
+			if len(claims.Tasks) > 0 {
+				lifecycle.ReportTaskClaimed(len(claims.Tasks))
+			}
+		}
+
+		// If we received zero claims or encountered an error, we wait at-least
+		// pollingInterval before polling again. We start the timer here, so it's
+		// counting while we wait for capacity to be available.
+		var delay <-chan time.Time
+		if claims == nil || len(claims.Tasks) == 0 {
+			delay = time.After(w.pollingInterval)
+		} else {
+			delay = time.After(0)
```

Comment:
I believe task claiming process could run in less time than `pollingInterval`, in which case, we could end up polling more frequently than `pollingInterval`. Also if multiple go routines were claiming work we'd be polling too frequently. Also this is code that would be easy to mess up, so any `0` delays are particularly dangerous as a bug could DDoS our services - would be better for `ClaimWork` to globally throttle, as a centralised safety net, and would also simplify code like this code above.